### PR TITLE
feature extractor

### PR DESCRIPTION
**Pull request by clewis7**
implementing training and inference using feature extractor

### CODE REVIEW

File: animal_soup/extensions/feature_extr_extensions.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        batch_size: int, default 32
+            Batch size.
+        gpu_id: int, default 0
+            Specify which gpu to use for training the model.
+        initial_lr: float, default 0.0001
+            Initial learning rate.
+        stop_method: str, default learning_rate
+            Method for stopping training. Argument must be one of ["early", "learning_rate", "num_epochs"]
+
+            | stop method   | description                                                                |
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:82]: "early" not in table

File: animal_soup/extensions/feature_extr_extensions.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            Location of checkpoint used for flow generator. If None, then will use default checkpoint of flow
+            generator.
+        model_in: str or Path, default None
+            If you want to retrain the model using different model weights than the default. User can
+            provide a location to a different model checkpoint. For example, if you had retrained the feature extractor
+            previously and wanted to use those weights instead.
+        model_out: str or Path, default None
+            User provided location of where to store model output such as model checkpoint with updated weights,
+            hdf5 file with model results/metrics, etc. Should be a directory. By default, the model output will get
+            stored in the same directory as the dataframe.
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:98]: maybe rename these to something like `"flow_model"`, `"feature_model_in"`, `"feature_model_out"`?

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: animal_soup/extensions/feature_extr_extensions.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        else:
+            df_path = self._df.paths.get_df_path()
+            df_dir, relative = self._df.paths.split(df_path)
+            os.makedirs(df_dir.joinpath("feature_extr_output"), exist_ok=True)
+            model_out = df_dir.joinpath("feature_extr_output")
+        if os.listdir(model_out):
+            raise ValueError(f"directory to store model output should be empty")
+
+        # validate only one experiment type being used in training
+        if len(list(set(list(self._df["exp_type"])))) != 1:
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:147]: wat, how did you end up with this O_O

Just use:

```python
if np.unique(df["exp_type"].values).size > 1
```

File: animal_soup/extensions/feature_extr_extensions.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+            raise ValueError("Training can only be completed with experiments of same type. "
+                             f"The current experiments in your dataframe are: {set(list(self._df['exp_type']))} "
+                             "Take a subset of your dataframe to train with one kind of experiment.")
+            # validate that an exp_type has been set
+        if list(set(list(self._df["exp_type"])))[0] is None:
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:152]: Is this to check if any row has `exp_type` == `None`? Should also check for nan. Anyways, if so this is not the right way to check.

Comment (kushalkolar) [reply animal_soup/extensions/feature_extr_extensions.py:152]: I think `if None in np.unique(df["exp_type"].values)` might work. 

If not then this will work:

```python
if None in np.unique(df["exp_type"].values).tolist()
```

File: animal_soup/extensions/feature_extr_extensions.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+
+        # check gpu_id
+        gpu_options = get_gpu_options()
+        if gpu_id not in gpu_options.keys():
+            raise ValueError(
+                f"gpu_id: {gpu_id} not in {gpu_options}. " f"Please select a valid gpu."
+            )
+
+        # check batch_size
+        if batch_size < MIN_BATCH_SIZE or batch_size > MAX_BATCH_SIZE:
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:169]: python lets you have expressive math operators :smile: , so you can just do:

```python
if not MIN_BATCH_SIZE < batch_size < MAX_BATCH_SIZE:
```

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: animal_soup/data/datasets.py [hunk 6]
```diff
@@ -205,19 +232,48 @@ def __init__(
...
-                vid_paths[i],
-                transform=transform,
-                conv_mode=conv_mode,
-                mean_by_channels=mean_by_channels,
-                frames_per_clip=frames_per_clip,
-            )
-            datasets.append(dataset)
-            dataset_info.append(dataset.metadata)
+
+        if labels is None: # non supervised training
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/data/datasets.py:241]: if you are doing flow generator training than that is unsupervised

File: animal_soup/data/utils.py [hunk 7]
```diff
@@ -9,12 +10,54 @@ def prepare_label(label: np.ndarray) -> np.ndarray:
     label = label.T
     # add background row
     rows = list()
-    for r in label:
-        if sum(r) > 0:
+    for row in label:
+        # sum across all behaviors for a time point
+        # insert in first position the background value
+        # for that time point
```
Comment (kushalkolar) [inline animal_soup/data/utils.py:16]: explain this better, sum across all behaviors at a given timepoint, if the sum is `0`, set `background = 1`

How fast is this when you have hundreds of ethograms? It's looks like a relatively slow for loop, but it might be more than sufficient for this use case.

File: animal_soup/data/datasets.py [hunk 8]
```diff
@@ -275,6 +275,8 @@ def __init__(
 
                 self.labels = np.concatenate(final_labels)
 
+                self.pos_weight = make_loss_weight(self.num_pos, self.num_neg)
```
Comment (kushalkolar) [inline animal_soup/data/datasets.py:278]: what's `pos_weight`? more useful name?

File: animal_soup/extensions/feature_extr_extensions.py [hunk 9]
```diff
@@ -109,12 +114,17 @@ def train(
         flow_model_in: str or Path, default None
             Location of checkpoint used for flow generator. If None, then will use default checkpoint of flow
             generator based on the mode.
+        flow_mode: str, default None
+            One of ["slow", "medium", "fast"]. If you are using a different flow generator checkpoint than the
+            default, you need to specify the mode so the correct flow generator and flow classifier can be
+            reconstructed.
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:120]: I don't understand what happens when the default is `None`?

File: animal_soup/extensions/feature_extr_extensions.py [hunk 10]
```diff
@@ -124,9 +134,22 @@ def train(
...
-        # validate flow_model_in
+        # validate flow_model_in and flow_mode
         if flow_model_in is not None:
             flow_model_in = validate_checkpoint_path(flow_model_in)
+            # if flow_mode is None raise, need to know what model to reconstruct
+            if flow_mode is None:
+                raise ValueError("If you are using a non-default flow generator model checkpoint, you must"
+                                 "also specify the corresponding mode for the model you are reconstructing."
+                                 "See below for mode/model correspondence: \n "
+                                 "{'slow': 'TinyMotionNet3D', 'medium': 'MotionNet', 'fast': 'TinyMotionNet}")
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:145]: the braces `{` are not necessary here in just an error message :smile:

File: animal_soup/extensions/feature_extr_extensions.py [hunk 11]
```diff
@@ -144,14 +167,15 @@ def train(
             raise ValueError(f"directory to store model output should be empty")
 
         # validate only one experiment type being used in training
-        if len(list(set(list(self._df["exp_type"])))) != 1:
+        if len(set(self._df["exp_type"].values)) > 1:
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:170]: does `np.unique()` not work here? It should since `.values` returns a 1d numpy array.

File: animal_soup/extensions/feature_extr_extensions.py [hunk 12]
```diff
@@ -377,7 +558,9 @@ def _build_classifier(
     Returns
     -------
     model
-        One of [ResNet18, ResNet50, ResNet34-3D]. Depends on mode.
+        One of [ResNet18, ResNet50, ResNet34-3D]. Depends on mode. Either a flow classifier or a spatial classifier.
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:561]: put mode inside double backticks so it's formats distinctly in the docs

File: animal_soup/extensions/feature_extr_extensions.py [hunk 13]
```diff
@@ -462,10 +645,52 @@ def _build_classifier(
...
+    Parameters
+    ----------
+    spatial_classifier
+        Spatial classifier model with pre-trained weights already loaded.
+        One of [resnet18, resnet50, resnet34-3D] depending on the mode.
+    flow_classifier
+        Flow classifier model with pre-trained weights already loaded.
+        One of [resnet18, resnet50, resnet34-3D] depending on the mode.
+    fusion_type: str
+        One of ["average", "weighted_average", "concatenate"]. Indicates the way that fusing the classifier
```
Comment (kushalkolar) [inline animal_soup/extensions/feature_extr_extensions.py:670]: this isn't necessary for the user since this is a private method, but I'm curious what's the difference between these fusion types?

File: animal_soup/feature_extractor/inference.py
Comment (kushalkolar) [inline animal_soup/feature_extractor/inference.py:1]: what's the point of this file?

File: animal_soup/utils/defaults.py
Comment (kushalkolar) [inline animal_soup/utils/defaults.py:1]: A dict of dicts like this gets very messy to read. I would make a yaml file and read that yaml in as a dict.

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: animal_soup/extensions/feature_extr_extensions.py (hunk 11 above)
Comment (kushalkolar) [reply animal_soup/extensions/feature_extr_extensions.py:173]: That's wird I'm curious, what did you get?

File: animal_soup/feature_extractor/inference.py [hunk 14]
```diff
@@ -2,131 +2,127 @@
...
+        with torch.no_grad():
+            images = gpu_transform(images)
+
+            logits = hidden_two_stream(images)
+            spatial_features = activation['spatial']
+            flow_features = activation['flow']
+        
+        # because we are using iterable datasets, each batch will be a consecutive chunk of frames from one worker
+        # but they might be from totally different chunks of the video. therefore, we return the frame numbers,
+        # and use this to store into our buffer in the right location
```
Comment (kushalkolar) [inline animal_soup/feature_extractor/inference.py:107]: can you come up with a test that checks that this works correctly?

File: animal_soup/extensions/feature_extr_extensions.py (hunk 11 above)
Comment (clewis7) [reply animal_soup/extensions/feature_extr_extensions.py:173]: part of `np.unique` is that it does some kind of ordering, and so if there is `None` in the unique values...it was giving error because it was trying to order a `str` and `None` type

File: animal_soup/data/datasets.py [hunk 15]
```diff
@@ -285,3 +286,167 @@ def __len__(self):
 
     def __getitem__(self, index: int):
         return self.dataset[index]
+
+
+# https://pytorch.org/docs/stable/data.html
+class VideoIterable(data.IterableDataset):
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/data/datasets.py:292]: the video files in DEG remain with whatever codec they originally were, the hdf5 files are for storing outputs not anything to do with the videos

File: animal_soup/extensions/feature_extr_extensions.py [hunk 16]
```diff
@@ -521,7 +522,30 @@ def infer(
...
+
+        prediction_info = predict_single_video(
+            vid_path=resolve_path(self._series["vid_path"]),
+            hidden_two_stream=hidden_two_stream,
+            fusion_type=DEFAULT_CLASSIFIER_AUGS["fusion"],
+            mean_by_channels=AUGS["normalization"]["mean"],
+            gpu_id=gpu_id,
+            flow_window=flow_window,
+            cpu_transform=get_cpu_transforms(AUGS),
+            gpu_transform=get_gpu_inference_transforms(AUGS, conv_mode=conv_mode)
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/feature_extr_extensions.py:545]: for gpu inferences you just convert to float and normalize, cpu transforms you just crop and resize

File: animal_soup/feature_extractor/inference.py [hunk 17]
```diff
@@ -2,131 +2,127 @@
...
+        One of ["average", "weighted_average", "concatenate"]. Indicates the way that fusing the classifier
+        models together should occur.
+    flow_window: int, default 11
+            Flow window size. Used to infer optic flow features to pass to the feature extractor.
+    mean_by_channels : np.ndarray
+        Image channel mean for z-scoring
+    cpu_transform : callable, optional
+        CPU transforms to perform, e.g. center cropping / resizing, by default None
+    gpu_transform : callable, optional
+        GPU augmentations. For inference, should just be conversion to float and z-scoring, by default None
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/feature_extractor/inference.py:44]: oh, shouldn't be optional...think that was left over doc string from deg I need to fix

File: animal_soup/feature_extractor/inference.py (hunk 14 above)
Comment (clewis7) [reply animal_soup/feature_extractor/inference.py:107]: yeah, I think next week I will spend some time working on testing

File: animal_soup/extensions/feature_extr_extensions.py (hunk 11 above)
Comment (kushalkolar) [reply animal_soup/extensions/feature_extr_extensions.py:173]: ah interesting thanks! Now I know of a usecase where `set` works but `np.unique` doesn't :smile: . So I guess in this case the `.values` returns an array with `dtype` `object` instead of `str` since it's mixed types.

File: animal_soup/feature_extractor/inference.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kushalkolar) [reply animal_soup/feature_extractor/inference.py:44]: ah so basically zscore is necessary for it to work properly I guess? makes sense because zscoring is often required preprocessing for any type of classification
